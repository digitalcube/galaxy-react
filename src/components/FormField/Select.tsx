import React, { FC, ReactNode } from 'react';
import { FaAngleDown } from 'react-icons/fa';
import { FormGroup, InputGroup, Label, Input, InputProps } from 'reactstrap';

export type FormSelectProps = {
  options: React.OptionHTMLAttributes<HTMLOptionElement>[];
  id: string;
  label: string;
  icon?: ReactNode;
  value: string;
  required?: boolean;
  readOnly?: boolean;
  disabled?: boolean;
  onChange: React.ChangeEventHandler<HTMLInputElement>;
  errorMessage?: string;
  inputOptions?: Omit<
    InputProps,
    | 'onChange'
    | 'type'
    | 'value'
    | 'id'
    | 'required'
    | 'readOnly'
    | 'disabled'
    | 'children'
  >;
  className?: string;
};
export const FormSelect: FC<FormSelectProps> = ({
  id,
  options,
  value,
  label,
  onChange,
  required,
  readOnly,
  disabled,
  errorMessage,
  inputOptions,
  className,
}) => {
  const inputClassName = [
    inputOptions?.className,
    !!errorMessage ? 'is-invalid' : '',
  ]
    .filter(Boolean)
    .join(' ');
  const inputProps: InputProps = {
    type: 'select',
    value,
    id,
    className: inputClassName,
    placeholder: label,
    onChange,
    ...inputOptions,
  };
  if (required) inputProps.required = required;
  if (disabled) inputProps.disabled = disabled;
  if (readOnly) inputProps.readOnly = readOnly;
  return (
    <FormGroup className={`mb-4${className ? ` ${className}`: ''}`}>
      <Label className="font-weight-bold" htmlFor={id}>
        {label}
      </Label>
      <InputGroup className="position-relative select-wrapper">
        <Input {...inputProps}>
          {options.map((optionProps, i) => (
            <option key={`${id}-option-${i}`} {...optionProps} />
          ))}
        </Input>
          <i className="position-absolute" aria-hidden="true">
            <FaAngleDown className="d-block" />
          </i>
        {errorMessage ? (
          <div className="invalid-feedback">{errorMessage}</div>
        ) : null}
      </InputGroup>
    </FormGroup>
  );
};
