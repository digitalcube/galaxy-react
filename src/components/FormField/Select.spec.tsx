import { render } from '@testing-library/react';
import React from 'react';
import { Default as FormSelect } from './Select.stories';

describe('FormSelect', () => {
  it('to match snapshot', () => {
    const { asFragment } = render(<FormSelect {...(FormSelect.args as any)} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('to match snapshot (with error massage)', () => {
    const { asFragment } = render(
      <FormSelect
        {...(FormSelect.args as any)}
        errorMessage="We donâ€™t recognize that email"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
