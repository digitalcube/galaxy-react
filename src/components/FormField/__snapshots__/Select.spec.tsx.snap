// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect to match snapshot (with error massage) 1`] = `
<DocumentFragment>
  <div
    class="mb-4 form-group"
  >
    <label
      class="font-weight-bold"
      for="password"
    >
      Password
    </label>
    <div
      class="input-group"
    >
      <select
        class="is-invalid form-control"
        id="password"
        placeholder="Password"
      >
        <option
          value="jp"
        >
          Japanese
        </option>
        <option
          value="en"
        >
          English
        </option>
      </select>
      <div
        class="invalid-feedback"
      >
        We donâ€™t recognize that email
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FormSelect to match snapshot 1`] = `
<DocumentFragment>
  <div
    class="mb-4 form-group"
  >
    <label
      class="font-weight-bold"
      for="password"
    >
      Password
    </label>
    <div
      class="input-group"
    >
      <select
        class="form-control"
        id="password"
        placeholder="Password"
      >
        <option
          value="jp"
        >
          Japanese
        </option>
        <option
          value="en"
        >
          English
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
